packages/server/src/migrations/functions/tableSettings.ts

```diff
@@ -0,0 +1,143 @@
+import { getScreenParams } from "../../db/utils"
+import { Screen } from "@budibase/types"
+import { makePropSafe as safe } from "@budibase/string-templates"
+/**
+ * Date:
+ * November 2022
+ *
+ * Description:
+ * Update table settings to use actions instead of links. We do not remove the
+ * legacy values here as we cannot guarantee that their apps are up-t-date.
+ * It is safe to simply save both the new and old structure in the definition.
+ *
+ * Migration 1:
+ * Legacy "linkRows", "linkURL", "linkPeek" and "linkColumn" settings on tables
+ * and table blocks are migrated into a "Navigate To" action under the new
+ * "onClick" setting.
+ *
+ * Migration 2:
+ * Legacy "titleButtonURL" and "titleButtonPeek" settings on table blocks are
+ * migrated into a "Navigate To" action under the new "onClickTitleButton"
+ * setting.
+ */
+export const run = async (appDb: any) => {
```

Comment:
A general requirement of migrations are that they are idempotent. Would it be safe if this migration was to be run twice on the same app?

It could happen where get through some screens successfully, but another screen fails. The entire migration would not succeed and would start again on all screens the next run.

Reply:
Yep - the migration will only change components that do not have the new settings defined. Once the migration has run and the new settings are defined, running the migration again would do nothing.